0000-macro-value.md

```diff
@@ -0,0 +1,300 @@
+- Feature Name: `macro_value`
+- Start Date: 2021-03-27
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a new macro_rules matcher, `$name:value`, with identical semantics to that of a function capture.
+
+# Motivation
+[motivation]: #motivation
+
+Value arguments to function-like macros are tricky to deal with.
+While macro_rules macros don't suffer from the most common pitfalls of C-style preprocessor macros,
+such as misnested brackets and operator precedence,
+using an `$:expr` capture more than once still evaluates the expression more than once,
+duplicating side effects.
+
+Additionally,
+we have the additional wrinkle of the lifetime and drop timing of temporaries complicating matters further,
+if your intent is to write a macro invocation with equivalent-to-function-call semantics.
+Suffice to say,
+`let arg = $arg;` has the incorrect drop behavior,
+and the current best practice is to instead expand to
+
+```rust
+match ( $arg0, $arg1, ) {
+    ( arg0, arg1, ) => { /* macro body */ }
+}
+```
+
+instead.
+We can simplify this and make getting the correct behavior easier on macro authors.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+(in a section explaining `macro_rules` matchers:)
+
+While `$:expr` is good for capturing an expression and copying that expression into the macro-expanded code,
+it does exactly that: it duplicates the captured expression to every expansion of the capture.
+If, for example, you wrote the trivial `min!` macro,
+
+```rust
+macro_rules! min {
+    ( $a:expr, $b:expr ) => {
+        if $a <= $b { $a } else { $b }
+    };
+}
+```
+
+then both `$a` and `$b` are evaluated twice,
+once at each expansion point,
+as opposed to a single time,
+as would be the case if [`min` were a function](https://doc.rust-lang.org/std/cmp/fn.min.html).
+If you want the arguments to the macro to be evaluated a single time,
+as if they were simple function arguments,
+you can use the `$:value` matcher:
+
+```rust
+macro_rules! min {
+    ( $a:value, $b:value ) => {
+        if $a <= $b { $a } else { $b }
+    };
+}
+```
+
+This time,
+`$a` and `$b` are evaluated a single time upon invoking the macro,
```

Comment:
This is confusing because macros are not "invoked" at runtime.

Reply:
Would
```suggestion
`$a` and `$b` are evaluated a single time when control flow enters the macro,
```
be clearer?

Reply:
"expanded macro" and then yes!